File: AgeOfArenaUnity/Assets/Scripts/FogOfWarSystem.cs
using UnityEngine;

/// <summary>
/// Classic RTS Fog of War — visual only (server-side AI/combat see everything).
///
/// A 128×128 CPU <see cref="Texture2D"/> covers the 120×120 world.  Each frame it
/// is repainted from team-0 sight circles; the result is uploaded to the GPU and
/// read by <see cref="FogOfWar"/> (Custom/FogOfWar) on the ground mesh.
/// Enemy unit and building renderers are toggled on/off every
/// <see cref="VisCheckInterval"/> seconds based on whether their grid cell is
/// currently lit.
///
/// Three visibility tiers (stored in the Red channel):
///   0   — unexplored (never seen):  ground is black.
///   70  — shroud (explored, not now visible): ground is dim.
///   255 — currently visible: ground at full colour.
/// </summary>
public class FogOfWarSystem : MonoBehaviour
{
    // ── Constants ─────────────────────────────────────────────────────────────
    const int   TexSize  = 128;
    const float WorldHalf = 60f;   // world is -60..+60 on X and Z
    const float WorldSize = WorldHalf * 2f;
    const float PixPerUnit = TexSize / WorldSize;  // pixels per world-unit (~1.067)

    const float VisCheckInterval = 0.5f;

    /// <summary>FOWD: Master switch — defaults to true (classic AoE2 fog). Set false to
    /// reveal the whole map (no fog); Init stays inert and Update no-ops.</summary>
    public bool fogEnabled = true;

    static readonly Color32 Black  = new Color32(0,   0,   0,   255);
    static readonly Color32 Shroud = new Color32(70,  70,  70,  255);
    static readonly Color32 Lit    = new Color32(255, 255, 255, 255);

    // ── State ─────────────────────────────────────────────────────────────────
    Texture2D _fogTex;

    /// <summary>MMTR: the fog texture (null if fog disabled). MinimapSystem reads this
    /// to overlay explored/unexplored areas on the minimap.</summary>
    public Texture2D FogTexture => fogEnabled ? _fogTex : null;
    Color32[] _pixels;
    byte[]    _explored;   // 0 = never seen, 1 = previously seen (shroud)

    float _visTimer;

    // ── Init / Reset ──────────────────────────────────────────────────────────

    /// <summary>
    /// Must be called once after the world is built.  Replaces the ground
    /// renderer's material with one that reads from the fog texture.
    /// </summary>
    public void Init(MeshRenderer groundRenderer)
    {
        // Fog disabled: leave the plain green ground in place and reveal everything.
        // Update() no-ops (guarded below), so enemy renderers are never hidden.
        if (!fogEnabled) return;

        _fogTex = new Texture2D(TexSize, TexSize, TextureFormat.RGB24, false)
        {
            filterMode = FilterMode.Bilinear,
            wrapMode   = TextureWrapMode.Clamp,
            name       = "FogOfWarTex",
        };
        _pixels   = new Color32[TexSize * TexSize];
        _explored = new byte[TexSize * TexSize];

        var fogMat = new Material(Shader.Find("Custom/FogOfWar"));
        fogMat.color = Prims.Hex(0x5b8c3e);            // match original ground green
        fogMat.SetTexture("_FogTex", _fogTex);
        groundRenderer.sharedMaterial = fogMat;
        groundRenderer.receiveShadows = true;

        // Start completely dark; the first Update() will reveal the TC area.
        Reset();
    }

    /// <summary>Clear explored state and repaint to black.  Called on Restart via
    /// <see cref="GameBootstrap"/> (the whole WorldRoot is rebuilt, so this system
    /// is recreated fresh — Reset() is here for manual/test use).</summary>
    public void Reset()
    {
        if (_explored == null) return;
        System.Array.Clear(_explored, 0, _explored.Length);
        for (int i = 0; i < _pixels.Length; i++) _pixels[i] = Black;
        _fogTex.SetPixels32(_pixels);
        _fogTex.Apply(false);
    }

    // ── Per-frame ─────────────────────────────────────────────────────────────

    void Update()
    {
        if (!fogEnabled) return;
        var gm = GameManager.Instance;
        if (gm == null || _fogTex == null) return;

        TickFogTexture(gm);

        _visTimer -= Time.deltaTime;
        if (_visTimer <= 0f)
        {
            _visTimer = VisCheckInterval;
            TickEnemyVisibility(gm);
        }
    }

    void TickFogTexture(GameManager gm)
    {
        // 1. Reset to shroud / unexplored
        for (int i = 0; i < _pixels.Length; i++)
            _pixels[i] = _explored[i] > 0 ? Shroud : Black;

        // 2. Paint sight circles for player (team 0) + allied teams
        for (int i = 0; i < gm.units.Count; i++)
        {
            var u = gm.units[i];
            if (u == null || !gm.IsAllied(0, u.teamId)) continue;
            // N10.minimap: elevation-aware sight — units on higher terrain see up to +20% farther.
            float elevBonus = 1f + Mathf.Clamp01(WorldRoot.GetHeight(
                u.transform.position.x, u.transform.position.z) / 1.4f) * 0.20f;
            PaintCircle(u.transform.position, UnitSight(u.type) * elevBonus);
        }

        // 3. Paint sight circles for player (team 0) + allied buildings
        for (int i = 0; i < gm.buildings.Count; i++)
        {
            var b = gm.buildings[i];
            if (b == null || !gm.IsAllied(0, b.teamId)) continue;
            PaintCircle(b.transform.position, BuildingSight(b.type));
        }

        // 4. Update persistent explored map
        for (int i = 0; i < _pixels.Length; i++)
            if (_pixels[i].r == 255) _explored[i] = 1;

        _fogTex.SetPixels32(_pixels);
        _fogTex.Apply(false);
    }

    /// <summary>Stamp a filled circle of <see cref="Lit"/> pixels at the given
    /// world position with the given sight radius (world units).</summary>
    void PaintCircle(Vector3 worldPos, float radiusU)
    {
        float rp  = radiusU * PixPerUnit;   // radius in pixels
        float rp2 = rp * rp;
        int   cx  = WorldToPixel(worldPos.x);
        int   cz  = WorldToPixel(worldPos.z);
        int   ext = Mathf.CeilToInt(rp) + 1;

        int x0 = Mathf.Max(0, cx - ext),   x1 = Mathf.Min(TexSize - 1, cx + ext);
        int z0 = Mathf.Max(0, cz - ext),   z1 = Mathf.Min(TexSize - 1, cz + ext);

        for (int z = z0; z <= z1; z++)
        {
            int rowOff = z * TexSize;
            float dz = z - cz;
            for (int x = x0; x <= x1; x++)
            {
                float dx = x - cx;
                if (dx * dx + dz * dz <= rp2)
                    _pixels[rowOff + x] = Lit;
            }
        }
    }

    // ── Enemy visibility ──────────────────────────────────────────────────────

    /// <summary>Show/hide enemy units and buildings based on whether their
    /// map cell is currently lit (r == 255).  Runs every 0.5 s to amortize cost.
    /// Allied units are always visible regardless of fog.</summary>
    void TickEnemyVisibility(GameManager gm)
    {
        for (int i = 0; i < gm.units.Count; i++)
        {
            var u = gm.units[i];
            if (u == null || gm.IsAllied(0, u.teamId)) continue;
            SetRenderersEnabled(u.gameObject, IsLit(u.transform.position));
        }
        for (int i = 0; i < gm.buildings.Count; i++)
        {
            var b = gm.buildings[i];
            if (b == null || gm.IsAllied(0, b.teamId)) continue;
            SetRenderersEnabled(b.gameObject, IsLit(b.transform.position));
        }
    }

    bool IsLit(Vector3 worldPos)
    {
        int x = Mathf.Clamp(WorldToPixel(worldPos.x), 0, TexSize - 1);
        int z = Mathf.Clamp(WorldToPixel(worldPos.z), 0, TexSize - 1);
        return _pixels[z * TexSize + x].r == 255;
    }

    static void SetRenderersEnabled(GameObject go, bool on)
    {
        var rs = go.GetComponentsInChildren<Renderer>();
        for (int i = 0; i < rs.Length; i++) rs[i].enabled = on;
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    static int WorldToPixel(float world)
        => Mathf.RoundToInt((world + WorldHalf) * PixPerUnit);

    static float UnitSight(UnitType t) => t switch
    {
        UnitType.Scout     => 13f, // recon unit: widest sight, reveals the most fog
        UnitType.Cavalry   => 9f,
        UnitType.Archer    => 8f,
        UnitType.Militia   => 7f,
        UnitType.Villager  => 5f,
        UnitType.Trebuchet => 4f,
        _                  => 5f,
    };

    static float BuildingSight(BuildingType t) => t switch
    {
        BuildingType.TownCenter                                               => 10f,
        BuildingType.Castle                                                   => 8f,
        BuildingType.Barracks or BuildingType.ArcheryRange or BuildingType.Stable => 7f,
        _                                                                     => 5f,
    };
}

File: web/src/game/FogOfWarSystem.ts
/**
 * FogOfWarSystem.ts — FogOfWarSystem.cs port.
 * 128×128 CPU grid covers the 120×120 world.
 * Three tiers: unexplored (black), shroud (dim), visible (clear).
 *
 * Ground fog: a large PlaneGeometry floating just above the terrain,
 * textured with a CanvasTexture updated from the fog grid — same approach
 * as Unity's Custom/FogOfWar shader on the ground mesh.
 *
 * Enemy entity hiding: enemy unit/building root.visible toggled each
 * VIS_INTERVAL seconds based on whether their grid cell is lit.
 */
import * as THREE from "three";
import { BuildingType, UnitType } from "../core/GameTypes";
import { diplomacy } from "../core/Diplomacy";
import type { Unit } from "./Unit";
import type { Building } from "./Building";

const TEX_SIZE    = 128;
const WORLD_HALF  = 90;  // covers BaseDistance=84 bases with margin
const WORLD_SIZE  = WORLD_HALF * 2;
const PIX_PER_UNIT = TEX_SIZE / WORLD_SIZE;

const REPAINT_INTERVAL = 0.2;
const VIS_INTERVAL     = 0.5;

const UNEXPLORED = 0;
const SHROUD     = 1;
const VISIBLE    = 2;

function unitSight(t: UnitType): number {
  switch (t) {
    case UnitType.Scout:     return 16;
    case UnitType.Cavalry:   return 10;
    case UnitType.Archer:    return 9;
    case UnitType.Militia:   return 8;
    case UnitType.Villager:  return 6;
    case UnitType.Trebuchet: return 5;
    default:                 return 6;
  }
}

function buildingSight(b: Building): number {
  let base: number;
  switch (b.buildingType) {
    case BuildingType.TownCenter:   base = 14; break;
    case BuildingType.Outpost:      base = 12; break;
    case BuildingType.Castle:       base = 10; break;
    case BuildingType.BombardTower:
    case BuildingType.WatchTower:
    case BuildingType.Barracks:
    case BuildingType.ArcheryRange:
    case BuildingType.Stable:       base = 8; break;
    default:                        base = 6; break;
  }
  return base + b.sightBonus;
}

function worldToPixel(w: number): number {
  return Math.round((w + WORLD_HALF) * PIX_PER_UNIT);
}

export class FogOfWarSystem {
  private readonly vis      = new Uint8Array(TEX_SIZE * TEX_SIZE);
  private readonly explored = new Uint8Array(TEX_SIZE * TEX_SIZE);

  private repaintTimer = 0;
  private visTimer     = 0;

  /** Canvas used as the Three.js fog texture source. */
  private readonly fogCanvas: HTMLCanvasElement;
  private readonly fogCtx: CanvasRenderingContext2D;
  private readonly imageData: ImageData;
  private readonly fogTexture: THREE.CanvasTexture;

  /** The fog plane mesh sitting just above the ground. */
  readonly mesh: THREE.Mesh;

  constructor(scene: THREE.Scene) {
    // ── Fog canvas texture ─────────────────────────────────────────────────
    this.fogCanvas = document.createElement("canvas");
    this.fogCanvas.width  = TEX_SIZE;
    this.fogCanvas.height = TEX_SIZE;
    this.fogCtx   = this.fogCanvas.getContext("2d")!;
    this.imageData = new ImageData(TEX_SIZE, TEX_SIZE);

    // Fill fully black initially
    const d = this.imageData.data;
    for (let i = 0; i < TEX_SIZE * TEX_SIZE; i++) {
      d[i * 4 + 3] = 255;
    }

    this.fogTexture = new THREE.CanvasTexture(this.fogCanvas);
    this.fogTexture.magFilter = THREE.LinearFilter;
    this.fogTexture.minFilter = THREE.LinearFilter;

    // ── Fog plane mesh ─────────────────────────────────────────────────────
    // Size matches the world bounds (180×180 = WORLD_HALF*2), placed at y=0.05.
    const geo = new THREE.PlaneGeometry(WORLD_SIZE, WORLD_SIZE);
    const mat = new THREE.MeshBasicMaterial({
      map: this.fogTexture,
      transparent: true,
      depthWrite: false,
      side: THREE.FrontSide,
    });
    this.mesh = new THREE.Mesh(geo, mat);
    this.mesh.rotation.x = -Math.PI / 2; // lay flat
    this.mesh.position.y = 0.05;
    this.mesh.renderOrder = 1;
    scene.add(this.mesh);
  }

  tick(units: Unit[], buildings: Building[], dt: number) {
    this.repaintTimer -= dt;
    if (this.repaintTimer <= 0) {
      this.repaintTimer = REPAINT_INTERVAL;
      this._repaintGrid(units, buildings);
      this._uploadTexture();
    }

    this.visTimer -= dt;
    if (this.visTimer <= 0) {
      this.visTimer = VIS_INTERVAL;
      this._updateEnemyVisibility(units, buildings);
    }
  }

  private _repaintGrid(units: Unit[], buildings: Building[]) {
    for (let i = 0; i < this.vis.length; i++) {
      this.vis[i] = this.explored[i] > 0 ? SHROUD : UNEXPLORED;
    }

    for (const u of units) {
      if (!u.alive || (u.teamId !== 0 && !diplomacy.isAlly(u.teamId, 0))) continue;
      this._paintCircle(u.pos.x, u.pos.z, unitSight(u.unitType));
    }
    for (const b of buildings) {
      if (!b.alive || (b.teamId !== 0 && !diplomacy.isAlly(b.teamId, 0))) continue;
      this._paintCircle(b.pos.x, b.pos.z, buildingSight(b));
    }

    for (let i = 0; i < this.vis.length; i++) {
      if (this.vis[i] === VISIBLE) this.explored[i] = 1;
    }
  }

  private _paintCircle(wx: number, wz: number, radiusU: number) {
    const rp  = radiusU * PIX_PER_UNIT;
    const rp2 = rp * rp;
    const cx  = worldToPixel(wx);
    const cz  = worldToPixel(wz);
    const ext = Math.ceil(rp) + 1;

    const x0 = Math.max(0, cx - ext), x1 = Math.min(TEX_SIZE - 1, cx + ext);
    const z0 = Math.max(0, cz - ext), z1 = Math.min(TEX_SIZE - 1, cz + ext);

    for (let z = z0; z <= z1; z++) {
      const rowOff = z * TEX_SIZE;
      const dz = z - cz;
      for (let x = x0; x <= x1; x++) {
        const dx = x - cx;
        if (dx * dx + dz * dz <= rp2) this.vis[rowOff + x] = VISIBLE;
      }
    }
  }

  private _uploadTexture() {
    const d = this.imageData.data;
    for (let pz = 0; pz < TEX_SIZE; pz++) {
      // Three.js PlaneGeometry UV: V=0 at top (world Z = +60), V=1 at bottom (Z = -60).
      // Our grid row 0 = world Z=-60. So mirror vertically when writing to ImageData.
      const srcRow = TEX_SIZE - 1 - pz;
      for (let px = 0; px < TEX_SIZE; px++) {
        const tier = this.vis[srcRow * TEX_SIZE + px];
        const off  = (pz * TEX_SIZE + px) * 4;
        if (tier === VISIBLE) {
          d[off + 3] = 0;   // fully transparent
        } else if (tier === SHROUD) {
          d[off]     = 0;
          d[off + 1] = 0;
          d[off + 2] = 0;
          d[off + 3] = 110; // explored-but-not-visible: dark grey
        } else {
          d[off]     = 0;
          d[off + 1] = 0;
          d[off + 2] = 0;
          d[off + 3] = 205; // unexplored: near-black, faint terrain silhouette visible
        }
      }
    }

    this.fogCtx.putImageData(this.imageData, 0, 0);
    this.fogTexture.needsUpdate = true;
  }

  private _updateEnemyVisibility(units: Unit[], buildings: Building[]) {
    for (const u of units) {
      if (u.teamId === 0 || diplomacy.isAlly(u.teamId, 0)) continue;
      u.root.visible = this._isLit(u.pos.x, u.pos.z);
    }
    for (const b of buildings) {
      if (b.teamId === 0 || diplomacy.isAlly(b.teamId, 0)) continue;
      b.root.visible = this._isLit(b.pos.x, b.pos.z);
    }
  }

  private _isLit(wx: number, wz: number): boolean {
    const x = Math.max(0, Math.min(TEX_SIZE - 1, worldToPixel(wx)));
    const z = Math.max(0, Math.min(TEX_SIZE - 1, worldToPixel(wz)));
    return this.vis[z * TEX_SIZE + x] === VISIBLE;
  }

  isVisible(wx: number, wz: number): boolean {
    return this._isLit(wx, wz);
  }

  /** Cheat: mark entire map as explored + visible, reveal all enemies. */
  revealAll(): void {
    this.vis.fill(VISIBLE);
    this.explored.fill(VISIBLE);
    this._uploadTexture();
  }
}
